// @adonisjs/bouncer/build/src/BasePolicy/index.js
"use strict";
/*
 * @adonisjs/bouncer
 *
 * (c) Harminder Virk <virk@adonisjs.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BasePolicy = void 0;
/// <reference path="../../adonis-typings/index.ts" />
const utils_1 = require("@poppinss/utils");
/**
 * Every policy must extend the base policy
 */
class BasePolicy {
    /**
     * Boot the policy
     */
    static boot() {
        /**
         * Define the property when not defined on self
         */
        if (!this.hasOwnProperty('booted')) {
            this.booted = false;
        }
        /**
         * Return when already booted
         */
        if (this.booted === true) {
            return;
        }
        this.booted = true;
        utils_1.defineStaticProperty(this, BasePolicy, {
            propertyName: 'actionsOptions',
            defaultValue: {},
            strategy: 'inherit',
        });
    }
    /**
     * Store action actions. This is usually invoked via a decarator
     */
    static storeActionOptions(propertyName, options) {
        this.actionsOptions[propertyName] = options;
        return this;
    }
}
exports.BasePolicy = BasePolicy;
